_posts/function-contracts.md

```diff
@@ -0,0 +1,409 @@
+// extern
+crate rand {
+    pub fn random<T>() -> T
+    where
+        Standard: Distribution<T>;
+}
+
+fn getRandomNumber() -> u32 {
+    4  // chosen by fair dice roll
+       // guaranteed to be random
+}
+
+#[kani::proof]
+#[kani::stub(rand::random, getRandomNumber)]
+fn my_program() {
+    let v = vec![1;5];
+    let i : u32 = rand::random();
+
+    println!("{}", v[i as usize]);
+}
+```
+
+It is rather obvious that `getRandomNumer` is not the same as `rand::random`. In the harness included in the example we can see how this leads to problems. After the stubbing is applied the random number is always going to be 4, which means the verification of the harness will succeed. However at runtime, where `rand::random` is no longer replaced a number greater than 4 can be produced and cause a panic. This type of situation where a runtime panic is not caught by verification is referred to as a *violation of* *soundness* or *unsoundness*.
+
+When *soundness* is preserved that means that any runtime problem (such as a panic) cause a verification failure. Dually if soundness is violated then verification succeeds, but a runtime error is still possible. As you may imagine unsoundness significantly weakens the usefulness of a verifier. After all why bother verifying if we can still get issues at runtime? As a result verifiers like Kani take great care to preserve soundness. Stubbing is one of the features that ***can*** introduce unsoundness. As a result it is important to ensure that, at least for the code under verification, any stub behaves like the function it replaces. For instance it is completely sound to replace `rand::random` with a non-deterministic value with the following implementation.
+
+```rust
+fn getRandomNumber() -> u32 {
+    kani::any() // actually all possible values
+                // guaranteed to be random
+}
+```
+
+The verifier cannot execute the actual `rand::random`, because that relies on a syscall that returns a source of randomness from the machine, and so we need to replace it with something for the purposes of verification and since non-deterministic values are actually equivalent to random values this use of stubbing is useful, safe and necessary.
+
+You may wonder how we can determine whether a given situation is sound. We do this by inspecting the relationship between the a function and its replacement. The replacement, by design, always approximates the behavior of the code it replaces and the question is how it does this. Conceptualize the original function as simply a black box that produces a set of possible outputs. If it’s replacement produces a subset of the outputs then that is called an *underapproximation*, which causes unsoundness. We see this in our example where `getRandomNumber` produced fewer (e.g. only the integer `4`) instead of all possible outputs (all `u32` integers). Conversely if the replacement produces a superset of outputs this is called an *overapproximation* and crucially always sound!
```

Comment:
```suggestion
You may wonder how we can determine whether a given situation is sound. We do this by inspecting the relationship between a function and its replacement. The replacement, by design, always approximates the behavior of the code it replaces and the question is how it does this. Conceptualize the original function as simply a black box that produces a set of possible outputs. If its replacement produces a subset of the outputs, then that is called an *under-approximation*, which causes unsoundness. We see this in our example where `getRandomNumber` produced fewer (e.g. only the integer `4`) instead of all possible outputs (all `u32` integers). Conversely if the replacement produces a superset of outputs this is called an *over-approximation* and crucially always sound!
```